[MediaRecorder] timecode provided in BlobEvent is incorrect

jyavenard opened the pull request:
#### ced7f5e2066e77bb308efb5274a69119421cfd81
<pre>
[MediaRecorder] timecode provided in BlobEvent is incorrect
<a href="https://bugs.webkit.org/show_bug.cgi?id=282320">https://bugs.webkit.org/show_bug.cgi?id=282320</a>
<a href="https://rdar.apple.com/138895100">rdar://138895100</a>

Reviewed by Anne van Kesteren.

The timecode returned was always of the time at which requestData() was called and not
&quot;For a MediaRecorder instance, the timecode in the first produced BlobEvent MUST contain 0.
Subsequent BlobEvent&apos;s timecode contain the difference of the timestamp of creation of the
first chunk in said BlobEvent and the timestamp of the first chunk of the first produced BlobEvent&quot;

imported wpt&apos;s MediaRecorder-blob-timecode.https.html to test this functionality.
The test exposed an issue in our implementation (which the test didn&apos;t deal with).
It sets a timeslice of 0 and assume that we would always get a monotonically increasing
timecode, but if we attempts to make 0ms long segment that may not always be the case.

As a workaround, and considering real-life usage, if a timeslice is set and no data was recorded at the time
the timer occurs, we will not emit `dataavailable` event. We also limit how small the timeslice can be to 100ms.

On platforms using AVFoundation, we make the minimum segment size to be 1s. Using a MP4 recorder calling requestData()
will return empty Blobs if called before 1s of data has been accumulated due to the structure of ISOBMFF (it needs to
write a complete samples table first). An AVFormatWriter will also errors with &quot;NotReady&quot; if we attempt to generate
too short segments.
With WebM data is written as received, but a new cluster will always be at least one second long.

We also hold off writing any samples until the AudioSampleBufferCompressor has started to return audible data to
ensure video and audio frames are already interleaved in monotonically increasing time order.

* LayoutTests/imported/w3c/web-platform-tests/mediacapture-record/MediaRecorder-blob-timecode.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/mediacapture-record/MediaRecorder-blob-timecode.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/mediacapture-record/w3c-import.log:
* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/Modules/mediarecorder/MediaRecorder.cpp:
(WebCore::MediaRecorder::MediaRecorder):
(WebCore::MediaRecorder::startRecording):
(WebCore::MediaRecorder::requestData):
(WebCore::MediaRecorder::requestDataInternal): Split method. Don&apos;t emit dataavailable event unless we actually have new data.
* Source/WebCore/Modules/mediarecorder/MediaRecorder.h: Set a minimum timeslice of 0.1s.
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp:
(WebCore::MediaRecorderPrivateEncoder::takeData):
(WebCore::MediaRecorderPrivateEncoder::enqueueCompressedAudioSampleBuffers):
(WebCore::MediaRecorderPrivateEncoder::flushToEndSegment): Return pair returnCode/time instead of setting a member variable.
Test exposed a case where we could have not written all available frames.
(WebCore::MediaRecorderPrivateEncoder::muxNextFrame):
(WebCore::MediaRecorderPrivateEncoder::fetchData):
(WebCore::MediaRecorderPrivateEncoder::flushPendingData):
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.h:

Canonical link: <a href="https://commits.webkit.org/286442@main">https://commits.webkit.org/286442@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/14773f709983a768189da0401435ec46367b8883

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/75973 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/55002 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/28872 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/80470 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/27239 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/78089 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/64144 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/3298 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/59572 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17728 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/79040 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/49452 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/65244 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39928 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/46852 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/22728 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/25566 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/67968 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/23065 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/81934 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/3342 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/2125 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/67798 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/3496 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65212 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67108 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16723 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11056 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/9178 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/3292 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/6098 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/3313 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/4251 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/3320 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp:
```diff
@@ -905,11 +905,12 @@ void MediaRecorderPrivateEncoder::fetchData(CompletionHandler<void(RefPtr<Fragme
         return flushPendingData(currentTime)->whenSettled(queueSingleton(), [protectedThis, this, completionHandler = WTFMove(completionHandler), currentTime]() mutable {
             assertIsCurrent(queueSingleton());
             RefPtr data = takeData();
```
Drive-by: since it's assumed data is non-null, would it be cleaner to assign to a Ref?

youennf commented on LayoutTests/http/wpt/mediarecorder/MediaRecorder-audio-timecode-expected.txt:
```diff
@@ -0,0 +1,3 @@
+
+PASS BlobEvent timecode increases
```
Can you import WPT mediacapture-record/MediaRecorder-blob-timecode.https.html instead?

annevk reviewed (approved):
Nice, still looks good (and better because more WPT!), but I suspect this is going to fail on the bots because the test lacks an -expected.txt file.

jyavenard committed:
[MediaRecorder] timecode provided in BlobEvent is incorrect
https://bugs.webkit.org/show_bug.cgi?id=282320
rdar://138895100

Reviewed by Anne van Kesteren.

The timecode returned was always of the time at which requestData() was called and not
"For a MediaRecorder instance, the timecode in the first produced BlobEvent MUST contain 0.
Subsequent BlobEvent's timecode contain the difference of the timestamp of creation of the
first chunk in said BlobEvent and the timestamp of the first chunk of the first produced BlobEvent"

imported wpt's MediaRecorder-blob-timecode.https.html to test this functionality.
The test exposed an issue in our implementation (which the test didn't deal with).
It sets a timeslice of 0 and assume that we would always get a monotonically increasing
timecode, but if we attempts to make 0ms long segment that may not always be the case.

As a workaround, and considering real-life usage, if a timeslice is set and no data was recorded at the time
the timer occurs, we will not emit `dataavailable` event. We also limit how small the timeslice can be to 100ms.

On platforms using AVFoundation, we make the minimum segment size to be 1s. Using a MP4 recorder calling requestData()
will return empty Blobs if called before 1s of data has been accumulated due to the structure of ISOBMFF (it needs to
write a complete samples table first). An AVFormatWriter will also errors with "NotReady" if we attempt to generate
too short segments.
With WebM data is written as received, but a new cluster will always be at least one second long.

We also hold off writing any samples until the AudioSampleBufferCompressor has started to return audible data to
ensure video and audio frames are already interleaved in monotonically increasing time order.

* LayoutTests/imported/w3c/web-platform-tests/mediacapture-record/MediaRecorder-blob-timecode.https-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/mediacapture-record/MediaRecorder-blob-timecode.https.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/mediacapture-record/w3c-import.log:
* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/Modules/mediarecorder/MediaRecorder.cpp:
(WebCore::MediaRecorder::MediaRecorder):
(WebCore::MediaRecorder::startRecording):
(WebCore::MediaRecorder::requestData):
(WebCore::MediaRecorder::requestDataInternal): Split method. Don't emit dataavailable event unless we actually have new data.
* Source/WebCore/Modules/mediarecorder/MediaRecorder.h: Set a minimum timeslice of 0.1s.
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.cpp:
(WebCore::MediaRecorderPrivateEncoder::takeData):
(WebCore::MediaRecorderPrivateEncoder::enqueueCompressedAudioSampleBuffers):
(WebCore::MediaRecorderPrivateEncoder::flushToEndSegment): Return pair returnCode/time instead of setting a member variable.
Test exposed a case where we could have not written all available frames.
(WebCore::MediaRecorderPrivateEncoder::muxNextFrame):
(WebCore::MediaRecorderPrivateEncoder::fetchData):
(WebCore::MediaRecorderPrivateEncoder::flushPendingData):
* Source/WebCore/platform/mediarecorder/MediaRecorderPrivateEncoder.h:

Canonical link: https://commits.webkit.org/286442@main
